Bug 1467480 - serve common references here

djmitche opened the pull request:
I added versions, too.

djmitche committed:
Bug 1467480 - add common schemas

djmitche committed:
Bug 1467480 - make manifest a valid schema, adjust some patterns and descriptions

petemoore commented on schemas/manifest-v2.yml:
```diff
@@ -40,34 +41,34 @@ properties:
                 description: Version of API, e.g. `v1`.
                 type: string
-                pattern: ^v[0-9][0-9\.\-]*$
+                pattern: ^v[0-9][0-9]*$
```
This could be a little restrictive, e.g. if someone somewhere forks. That said, they could also fork the pattern, just a little more work for them.
If we intend to have 'v' + positive integer, maybe we should make the version an integer, with `minimum: 1` instead. Then taskcluster-lib-url could insert the `v` char, and the version property could just be an integer?

djmitche replied:
I don't think we'll be able to support forks.  Well, more like, I don't think we could support joins :)  We've never made a v2, and I think the cases where we might do so would be major platform changes, rather than something a user might do locally.  Just like a company wouldn't invent IPv7 internally.

We chose to include the "v" in the string here because that's how it appears in the API paths, schema paths, and reference paths, and because we typically refer to it as "vee one".  It helps to distinguish from metaschema version, library versions, service versions, etc.

jhford replied:
Using the v prefix here doesn't really solve for ambiguity though, since lots of things are tagged with a v prefix, including the the filename of this document, library versions and service versions.  I don't have any preference for having a v prefix or not in the schema, but I think Pete is right here, in that it's easier to add a v to integers where we consume the value than it is to split off a v when we're trying to read in the reference file.

imbstack replied:
I just feel like we call the api version `v1` in the api itself so it may as well be `v1` here too? I get the concern with blocking in forks. I expect that is a hill we can climb when we come to it as well. This does not lock us in forever!

petemoore commented on src/main.js:
```diff
@@ -4,7 +4,7 @@ const {store} = require('./store');
 const build = async (input, output, rootUrl) => {
   const {references, schemas} = await load({input});
-  update({references, schemas, rootUrl});
+  await update({references, schemas, rootUrl});
```
302 @jhford (I don't know ins and outs of `await`)

djmitche replied:
The `update` function became asynchronous in this patch (because it now calls async functions), so it must be called with `await`.  A downside of async in JS and Python is that it's easy to forget to add `await` yet see tests still pass.  But I remembered it this time :)

petemoore reviewed (approved):
The changes seem reasonable, but I'd like a node guru to take look too if possible. @jhford ?

jhford reviewed (commented):
Why are we creating new versions of schemas?  The only change that I see as possibly breaking is the version number pattern, all other things are just naming or description, which shouldn't break anything.  Having multiple versions of the schemas is going to be a decent amount of maintenance overhead.

djmitche commented:
It's relatively rare to change the schemas, but we have done it before -- we just added kind=hook to the actions schema, the manifest schema was completely rewritten, and we've added fields to exchanges and APIs.

Historically, we've happily changed schemas in-place.  That is, the JSON you fetch at a particular URL tomorrow is different from what you fetched yesterday.  We've gotten away with that, mostly because we don't validate everything, a little because we tend to keep copies of schemas around, rather than downloading them, and partly because the changes we've made have been forward-compatible.

So the version numbers are there with the intent of not playing fast-and-loose like this anymore.  If we decide to add another field to actions, for example, we'll define `action-schema-v3.yml` by copy-pasting from v2 and making the change.  Then existing `actions.json` files will continue to be valid against v2, and new can be validated against v3.

That's all I want to accomplish here: name schemas such that the schema at a particular URL doesn't ever change.

How to find the version of a particular file is something we can address later.  In checking, though, I did notice that actions have a `version` property and it's set to `1` so I will rename it v2 to `actions-schema-v1.yml` and leave the old version out.  api-reference.json has version: 0, so I'll rename that too.  And exhcanges-reference.json has version 0 or, bizarrely, 0.2.0.  All of the existing events.json files have version: 0.  So I'll rename that to v0 as well.

I'll also decode the JSON to YAML.  Those files were JSON in the source repos, but there's no reason to continue that :)

djmitche commented:
The manifest format was completely overhauled, so I think it makes sense to refer to it as v2.

All of these files should probably have a `$schema` instead of just `version: ..` but that can be done in a followup (@jhford are you interested in working on that?)

djmitche committed:
Bug 1467480 - align vN with the 'version' property, everything in YAML

djmitche commented:
@jhford ping?
